### ISSUE TITLE
ArrayType iterable $haystack

### ISSUE THREAD

**Issue by simPod** [consistence/consistence#42]
I fancy `ArrayType::mapByCallback`. However, I miss the possibility to pas `iterable` type as haystack instead of `array`. I undestand that ARRAYtype vs ITERABLE is a bit weird, not sure how to approach it. 

Does it make sense to change the $haystack's typehint in this library to `iterable`? Sounds a bit hacky to me but I though I'd ask anyway, maybe you might have better idea. Currently, I just reimplemented it myself in my codebase.

**Comment by VasekPurchart** [comment]
I agree with you, however I have a slightly different plan in mind - I am planning to have `IterableType` which will have all suitable equivalent methods.

**Comment by simPod** [comment]
Ok, identical approach here 👍 

![image](https://user-images.githubusercontent.com/327717/57524385-f00fd900-7327-11e9-96a7-6cc7985650b9.png)

We can leave this open then I guess (?)
